server/v2/stf/stf.go

```diff
@@ -108,17 +104,48 @@ func (s STF[T]) validateTx(ctx context.Context, state store.WritableState, gasLi
 	return validateCtx.gasUsed, validateCtx.events, applyStateChanges(state, validateState)
 }
 
+// execTx executes the tx messages on the provided state. If the tx fails then the state is discarded.
 func (s STF[T]) execTx(ctx context.Context, state store.WritableState, gasLimit uint64, tx T) (msgResp Type, gasUsed uint64, execEvents []event.Event, err error) {
 	execState := s.branch(state)
 	execCtx := s.makeContext(ctx, tx.GetSenders(), execState, gasLimit)
 	// atomic execution of the all messages in a transaction, TODO: we should allow messages to fail in a specific mode
-	for _, msg := range tx.GetMessages() {
-		msgResp, err = s.handleMsg(execCtx, msg)
+	var txErr error
+	for i, msg := range tx.GetMessages() {
+		msgResp, txErr = s.handleMsg(execCtx, msg)
 		if err != nil {
-			return nil, 0, nil, err
+			err = fmt.Errorf("tx execution failed at message with index %d: %w", i, err)
+			break // stop execution when one message fails.
+		}
+	}
+	// if tx failed then we run the post tx handler with the initial state, and we only return
+	// post tx handler events.
+	if txErr != nil {
```

Comment:
in the case of gas refunds, if execution makes it 50% of the way through and then errors, would we want to refund 50% or nothing?

Reply:
ah whoops read the flow wrong. still waking up sorry